[RPC] Give RPC commands more information about the RPC request

jonasschnelli opened the pull request:
Alternative solution for #8775.
Looks more invasive than it is (mostly search & replace).

This PR does replace the RPC call signature `(UniValue &params, bool fHelp)` with `(JSONRequest &request)` which allows the registered RPC calls to get more information about the request.

The `JSONRequest` object contains the HTTP base auth username as well as the path of the called HTTP endpoint.

This PR would allow wallet distinction based on HTTP auth username and or URL endpoints.

jonasschnelli committed:
[RPC] Give RPC commands more information about the RPC request

jonasschnelli committed:
[RPC] pass HTTP basic authentication username to the JSONRequest object

jonasschnelli commented:
For the multiwallet support, each wallet RPC call could do something like `CWallet *pwallet = CWallets::getWalletFromRequest(request)` instead of accessing `pwalletMain`.

laanwj commented:
Concept ACK, the code churn is a bit unfortunate but passing a context structure into RPC methods had to be done at some point...

jonasschnelli commented:
Yes. It's larger then I initially though. But once we did the "large" migration to a structure/object-passing we will be flexible for further changes.
Most changed lines are just a `params.` to `request.params.` which should be easy to verify/review.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -436,15 +436,15 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
+UniValue getmempoolancestors(const JSONRequest& request)
 {
-    if (fHelp || params.size() < 1 || params.size() > 2) {
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2) {
```
nit: You can put something like

```cpp
UniValue params = request.params;
```

in the first line of the method to make the diff smaller and less verbose?

jonasschnelli replied:
I have considered this. But if we would have done this from the beginning, wouldn't it be without the extra `UniValue params` delegation?
I think this PRs solution is cleaner.

We have to do this once and therefore probably accept a larger diff.

laanwj replied:
No strong opinion on this. Indeed, doing `const UniValue &params = request.params;` at the beginning of each function would reduce diff, on the other hand, I don't think prefixing `request.` everywhere is bad. It even makes the code a little bit more readable. "What params? The request params".
I'm fine with keeping it like this

laanwj commented:
This reminds me, I should have called that class `JSONRPCRequest` not `JSONRequest`. It's a JSON-RPC server not a JSON server.

If you agree I think it would make sense to rename it as a first commit (it only appears in 5 places) before it becomes an argument to every RPC call.

laanwj committed:
rpc: Change JSONRPCRequest to JSONRPCRequestObj

This is more consistent with `JSONRPCReplyObj`.

laanwj commented:
@jonasschnelli I'd suggest the following to get rid of the current `JSONRPCRequest`: https://github.com/laanwj/bitcoin/commit/ae6db2a5ff781ffa876f7d424438901d68529a8b, then rename the actual request object.

jonasschnelli commented:
Rebased and added https://github.com/laanwj/bitcoin/commit/ae6db2a5ff781ffa876f7d424438901d68529a8b as first commit, renamed `JSONRequest` to `JSONRPCRequest`.

laanwj commented:
utACK e7156ad
